src/internal/FluencePeer.ts

```diff
@@ -0,0 +1,350 @@
+
+    private _connection: FluenceConnection;
+    private _interpreter: AirInterpreter;
+
+    private async _initAirInterpreter(logLevel: AvmLoglevel): Promise<void> {
+        this._interpreter = await createInterpreter(this._interpreterCallback.bind(this), this._selfPeerId, logLevel);
+    }
+
+    private async _connect(multiaddr: MA, options?: FluenceConnectionOptions): Promise<void> {
+        const nodePeerId = multiaddr.getPeerId();
+        if (!nodePeerId) {
+            throw Error("'multiaddr' did not contain a valid peer id");
+        }
+
+        if (this._connection) {
+            await this._connection.disconnect();
+        }
+
+        const node = PeerId.createFromB58String(nodePeerId);
+        const connection = new FluenceConnection(
+            multiaddr,
+            node,
+            this._selfPeerIdFull,
+            this._executeIncomingParticle.bind(this),
+        );
+        await connection.connect(options);
+        this._connection = connection;
+        this._initWatchDog();
+    }
+
+    private async _disconnect(): Promise<void> {
+        if (this._connection) {
+            await this._connection.disconnect();
+        }
+        this._clearWathcDog();
+        this._requests.forEach((r) => {
+            r.cancel();
+        });
+    }
+
+    private get _selfPeerId(): PeerIdB58 {
+        return this._selfPeerIdFull.toB58String();
+    }
+
+    private get _relayPeerId(): PeerIdB58 | undefined {
+        return this._connection?.nodePeerId.toB58String();
+    }
+
+    private async _executeIncomingParticle(particle: Particle) {
+        logParticle(log.debug, 'external particle received', particle);
+
+        let request = this._requests.get(particle.id);
+        if (request) {
+            await request.receiveUpdate(particle);
+        } else {
+            request = RequestFlow.createExternal(particle);
+            request.handler.combineWith(this._callServiceHandler);
+        }
+        this._requests.set(request.id, request);
+
+        await this._processRequest(request);
+    }
+
+    private _processRequest(request: RequestFlow) {
+        try {
+            this._currentRequestId = request.id;
+            request.execute(this._interpreter, this._connection, this._relayPeerId);
+        } catch (err) {
+            log.error('particle processing failed: ' + err);
+        } finally {
+            this._currentRequestId = null;
+        }
+    }
+
+    private _interpreterCallback: ParticleHandler = (
+        serviceId: string,
+        fnName: string,
+        args: any[],
+        tetraplets: SecurityTetraplet[][],
+    ): CallServiceResult => {
+        if (this._currentRequestId === null) {
+            throw Error('current request can`t be null here');
+        }
+
+        const request = this._requests.get(this._currentRequestId);
+        const particle = request.getParticle();
+        if (particle === null) {
+            throw new Error("particle can't be null here");
```

Comment:
Let's add some debug info to the error? At least `this._currentRequestId`

Reply:
fixed